### PR TITLE
add FIM dataset support

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

Adds FIM dataset support.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/datasets/fim_dataset.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            sample = np.concatenate([sample, np.full((-1 * diff), self.pad_tok_id)])
+
+        assert sample.shape[0] == sample_len
+
+        return (
+            np.array(sample, dtype=np.int64),
+            np.array(document_ids, dtype=np.int64),
+        )
+
+    def _fim_permute_sequence(self, sequence, np_rng, rate):
```
Comment (suiyoubi) [inline megatron/training/datasets/fim_dataset.py]: why do we need to pass `np_rng` here ? could we directly infer from `self.np_rng` ?

File: megatron/training/datasets/fim_dataset.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            self.config.tokenizer,
+            truncate_or_pad=False,
+            suffix_tok_id=self.suffix_tok_id,
+            prefix_tok_id=self.prefix_tok_id,
+            middle_tok_id=self.middle_tok_id,
+            pad_tok_id=self.pad_tok_id,
+            no_fim_prefix=self.no_fim_prefix,
+        )
+
+    def _fim_split_and_permute_sequence(self, sequence, np_rng):
```
Comment (suiyoubi) [inline megatron/training/datasets/fim_dataset.py]: why do we need to pass np_rng here ? could we directly infer from self.np_rng ?

File: megatron/training/datasets/fim_dataset.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        # Permute the segment after the last split token
+        permuted = self._fim_permute_sequence(sequence[curr_start_position:], np_rng, self.fragment_fim_rate)
+        new_samples.append(permuted)
+
+        return np.concatenate(new_samples)
+
+    def _permute(
+        self,
+        sample,
+        np_rng,
```
Comment (suiyoubi) [inline megatron/training/datasets/fim_dataset.py]: why do we need to pass np_rng here ? could we directly infer from self.np_rng ?

File: tests/unit_tests/data/test_fim_dataset.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import torch
+
+from megatron.core.datasets.blended_megatron_dataset_builder import BlendedMegatronDatasetBuilder
+from megatron.core.datasets.utils import compile_helpers, get_blend_from_list
+from megatron.core.tokenizers import MegatronTokenizer
+from megatron.training.datasets.fim_dataset import GPTFIMDataset, GPTFIMDatasetConfig
+from tests.unit_tests.test_utilities import Utils
+
+
+def test_fim_gpt_dataset():
```
Comment (suiyoubi) [inline tests/unit_tests/data/test_fim_dataset.py]: I don't think this unit test is meaningful: For FIM we should test that we can detect that split: we should use the data in a way that mimic the FIM split: (e.g., set `rate=1.` and treat some tokens as "fake fim tokens" to force the FIM split )

File: megatron/training/datasets/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# Data Pipeline
+
+## FIM dataset
```
Comment (suiyoubi) [inline megatron/training/datasets/README.md:3]: Maybe not necessarily belong to this PR, but we should at least note in the readme that in order to use FIM training, your pretrain dataset needs to be preprocessed with the special tokens. We might need to add support for data preprocessing script as well ( could be separate PR)

File: megatron/training/datasets/fim_dataset.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        indexed_dataset: IndexedDataset,
+        dataset_path: str,
+        indexed_indices: np.ndarray,
+        num_samples: int,
+        index_split: Split,
+        config: GPTFIMDatasetConfig,
+    ) -> None:
+        super().__init__(indexed_dataset, dataset_path, indexed_indices, num_samples, index_split, config)
+
+        self.indexed_dataset = indexed_dataset
```
Comment (asolergi-nv) [inline megatron/training/datasets/fim_dataset.py]: `self.dataset` is defined already

https://github.com/NVIDIA/Megatron-LM/blob/00884a8765f2e67f880d962eaf3f231ed928fee2/megatron/core/datasets/megatron_dataset.py#L50

File: megatron/core/tokenizers/text/libraries/null_tokenizer.py [hunk 7]
```diff
@@ -25,6 +25,14 @@ def ids_to_text(self, ids):
         text = [str(x) for x in ids]
         return ' '.join(text)
 
+    def tokens_to_ids(self, tokens):
+        """Converts tokens to ids."""
+        return [int(x) for x in tokens]
```
Comment (asolergi-nv) [inline megatron/core/tokenizers/text/libraries/null_tokenizer.py:30]: This will raise a issue when trying to convert `str` to `int`. 

If we just want to test this feature with the `NullTokenizer` works since the `extra_tokens` you defined are already numbers. I would suggest to properly support this method & also test other tokenizers

File: megatron/training/datasets/fim_dataset.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+                # Add the sample part
+                offset = 0 if i > doc_index_beg else doc_index_beg_offset
+                length = None if i < doc_index_end else doc_index_end_offset + 1
+                sample_parts.append(self.indexed_dataset.get(self.document_index[i], offset=offset, length=length))
+
+        sample = np.concatenate(sample_parts)
+
+        sample_len = sample.shape[0]
+        segment_breaks = np.argwhere(sample == self.eod_tok_id)
+        np_rng = self.np_rng
```
Comment (asolergi-nv) [inline megatron/training/datasets/fim_dataset.py]:
```suggestion
```

File: megatron/training/datasets/fim_dataset.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        if doc_index_beg == doc_index_end:
+            # Add the document id
+            document_ids.append(self.document_index[doc_index_beg])
+
+            # Add the entire sample
+            sample_parts.append(
+                self.indexed_dataset.get(
+                    self.document_index[doc_index_beg],
+                    offset=doc_index_beg_offset,
+                    length=doc_index_end_offset - doc_index_beg_offset + 1,
```
Comment (asolergi-nv) [inline megatron/training/datasets/fim_dataset.py:133]:
```suggestion
                    length=doc_index_end_offset - doc_index_beg_offset + self.config.add_extra_token_to_sequence,
```

File: megatron/training/datasets/fim_dataset.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+
+        # Sample spans multiple documents
+        else:
+            for i in range(doc_index_beg, doc_index_end + 1):
+                # Add the document id
+                document_ids.append(self.document_index[i])
+
+                # Add the sample part
+                offset = 0 if i > doc_index_beg else doc_index_beg_offset
+                length = None if i < doc_index_end else doc_index_end_offset + 1
```
Comment (asolergi-nv) [inline megatron/training/datasets/fim_dataset.py:145]:
```suggestion
                length = None if i < doc_index_end else doc_index_end_offset + self.config.add_extra_token_to_sequence
```

File: megatron/core/tokenizers/text/libraries/null_tokenizer.py (hunk 7 above)
Comment (dimapihtar) [reply megatron/core/tokenizers/text/libraries/null_tokenizer.py:30]: > This will raise a issue when trying to convert `str` to `int`.
> 
> If we just want to test this feature with the `NullTokenizer` works since the `extra_tokens` you defined are already numbers. I would suggest to properly support this method & also test other tokenizers

`Null Tokenizer` doesn't support strings. That's a mock tokenizer for tests/debugging which takes only strings of integers.

File: megatron/training/datasets/fim_dataset.py (hunk 10 above)
Comment (dimapihtar) [reply megatron/training/datasets/fim_dataset.py:145]: I don't think we want to introduce new config value here since this extra token is expected to be added by default.

File: megatron/training/datasets/fim_dataset.py (hunk 9 above)
Comment (dimapihtar) [reply megatron/training/datasets/fim_dataset.py:133]: I don't think we want to introduce new config value here since this extra token is expected to be added by default.

File: megatron/training/datasets/README.md (hunk 5 above)
Comment (dimapihtar) [reply megatron/training/datasets/README.md:3]: agree, let's add it later i na separate PR.

File: megatron/training/datasets/fim_dataset.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+            # we should rather truncate at the context-level
+            if truncate_or_pad:
+                # need to make same length as the input. Take the 3 sentinel tokens into account
+                new_length = suffix.shape[0] + prefix.shape[0] + middle.shape[0] + 3
+                diff = new_length - sample.shape[0]
+                if diff > 0:  # too long
+                    if (
+                        suffix.shape[0] <= diff
+                    ):  # if there's no space to truncate the suffix: stop and report it. atm i should have stopped this from happening
+                        return sample, self.np_rng
```
Comment (suiyoubi) [inline megatron/training/datasets/fim_dataset.py]: this return does not seem right to me, should we just do `return sample`?

File: megatron/training/datasets/fim_dataset.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        """
+        If self.fim_split_sample is not None, split the sequence.
+        Then apply FIM on the fragments, or the whole sequence if self.fim_split_sample is None.
+        """
+        if self.fim_split_sample is None:
+            return self._fim_permute_sequence(sequence, self.fim_rate)
+        # fim_split_sample is set: split the sample on this token and permute each fragment separately.
+        # Typically, if each sample is a repository, then we split again on the file level.
+        # Each fragment is a file, and we permute the files.
+        fragment_breaks = np.argwhere(sequence == self.fim_split_sample)
```
Comment (suiyoubi) [inline megatron/training/datasets/fim_dataset.py:208]: Let's `assert` to make sure that `self.fim_split_sample` is not a vector (I remember we had some issue with starcoder for this) . We can do the assertion in the init func, something like this: 
```
ids = tokenizer.tokens_to_ids(split_sample)
assert isinstance(ids, int) or len(ids) == 1
self.fim_split_sample = ids if isinstance(ids, int) else ids[0]
```

File: megatron/training/datasets/README.md (hunk 5 above)
Comment (suiyoubi) [reply megatron/training/datasets/README.md:3]: yeah something like: 

```
fim_tokens["prefix"], fim_tokens["middle"], fim_tokens["suffix"], fim_tokens["pad"], fim_tokens["eod"]

These tokens must exist in the tokenizer vocab.

The dataset must be pre-tokenized with those tokens.
```

File: pretrain_gpt.py [hunk 13]
```diff
@@ -172,26 +173,49 @@ def core_gpt_dataset_config_from_args(args):
...
+        "allow_ambiguous_pad_tokens": args.allow_ambiguous_pad_tokens,
+    }
+
+    # add FIM args to the config
+    if args.fim_data:
+        extra_tokens = {
+            "prefix": args.fim_prefix_token,
+            "middle": args.fim_middle_token,
+            "suffix": args.fim_suffix_token,
+            "pad": args.fim_pad_token,
```
Comment (yaoyu-33) [inline pretrain_gpt.py:203]: will these be optional to add? will fim always has a different pad / eod token than original pad/ eod? and it's always required?

File: megatron/training/datasets/README.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+**Attributes**
+
+- `rate`: Probability of converting a sample into a FIM example.
+- `spm_rate`: Probability of using the SPM FIM pattern (vs PSM).
+- `extra_tokens`: Dictionary containing the FIM special tokens: {"prefix", "middle", "suffix", "pad", "eod"}.
+- `split_sample`: Optional token around which samples are split before applying FIM.
+- `fragment_rate`: Probability of applying FIM to each fragment when split_sample is used.
+- `no_prefix`: If the decoded sequence starts with this prefix, FIM is skipped.
+`GPTFIMDataset` dataset class that loads token sequences from an `IndexedDataset` and applies FIM transformations before returning each sample.
```
Comment (yaoyu-33) [inline megatron/training/datasets/README.md:19]: how will seq length change in this case? in many cases IndexedDataset has constant seq length?

File: megatron/training/datasets/README.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+`GPTFIMDataset` extends Megatron-Core’s `GPTDataset` to support **Fill-in-the-Middle (FIM)** data augmentation.
+It probabilistically converts samples into FIM format using configurable rates, with support for both PSM and SPM patterns, fragment-level splitting, and length-preserving output.
+
+`GPTFIMDatasetConfig` provides the configuration needed to enable this behavior.
+`GPTFIMDatasetConfig` configuration object extending `GPTDatasetConfig` to enable FIM preprocessing.
+
+**Attributes**
+
+- `rate`: Probability of converting a sample into a FIM example.
+- `spm_rate`: Probability of using the SPM FIM pattern (vs PSM).
```
Comment (yaoyu-33) [inline megatron/training/datasets/README.md]: what does this mean? 0.3 means 30% spm and 70% psm? be more clear.

File: megatron/training/datasets/README.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+- `extra_tokens`: Dictionary containing the FIM special tokens: {"prefix", "middle", "suffix", "pad", "eod"}.
+- `split_sample`: Optional token around which samples are split before applying FIM.
```
Comment (yaoyu-33) [inline megatron/training/datasets/README.md]: can you add an example, not fully understand.

File: megatron/training/datasets/README.md [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+- `fragment_rate`: Probability of applying FIM to each fragment when split_sample is used.
+- `no_prefix`: If the decoded sequence starts with this prefix, FIM is skipped.
```
Comment (yaoyu-33) [inline megatron/training/datasets/README.md:18]: `fim_skip_prefix` is better?

File: megatron/training/datasets/fim_dataset.py [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+from megatron.core.datasets.utils import Split
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class GPTFIMDatasetConfig(GPTDatasetConfig):
+    """Configuration object for Megatron Core GPT FIM datasets"""
+
+    rate: float = None
```
Comment (yaoyu-33) [inline megatron/training/datasets/fim_dataset.py]: i would add `fim_` prefix to configs, so when debugging and print all configs it's more clear what it is.

File: megatron/training/datasets/README.md (hunk 14 above)
Comment (dimapihtar) [reply megatron/training/datasets/README.md:19]: seq length will not be changed

File: pretrain_gpt.py (hunk 13 above)
Comment (dimapihtar) [reply pretrain_gpt.py:203]: not always, but it can be different to original pad / eod. It's up to user.

File: megatron/training/datasets/README.md (hunk 17 above)
Comment (dimapihtar) [reply megatron/training/datasets/README.md:18]: I'd leave it as it is since it's implemntation from NeMo1
